src/H5FDint.c

```diff
@@ -260,148 +300,2075 @@ H5FD_write(H5FD_t *file, H5FD_mem_t type, haddr_t addr, size_t size, const void
+                    /* Check if we're using the static arrays */
+                    if (addrs == addrs_static) {
+                        HDassert(sizes == sizes_static);
+                        HDassert(vec_bufs == vec_bufs_static);
+
+                        /* Allocate dynamic arrays */
+                        if (NULL == (addrs = H5MM_malloc(sizeof(addrs_static) * 2)))
+                            HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
+                                        "memory allocation failed for address list")
+                        if (NULL == (sizes = H5MM_malloc(sizeof(sizes_static) * 2)))
+                            HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
+                                        "memory allocation failed for size list")
+                        if (NULL == (vec_bufs = H5MM_malloc(sizeof(vec_bufs_static) * 2)))
+                            HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
+                                        "memory allocation failed for buffer list")
+
+                        /* Copy the existing data */
+                        (void)H5MM_memcpy(addrs, addrs_static, sizeof(addrs_static));
+                        (void)H5MM_memcpy(sizes, sizes_static, sizeof(sizes_static));
+                        (void)H5MM_memcpy(vec_bufs, vec_bufs_static, sizeof(vec_bufs_static));
+                    }
+                    else {
+                        void *tmp_ptr;
+
+                        /* Reallocate arrays */
+                        if (NULL == (tmp_ptr = H5MM_realloc(addrs, vec_arr_nalloc * sizeof(*addrs) * 2)))
+                            HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
+                                        "memory reallocation failed for address list")
+                        addrs = tmp_ptr;
+                        if (NULL == (tmp_ptr = H5MM_realloc(sizes, vec_arr_nalloc * sizeof(*sizes) * 2)))
+                            HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
+                                        "memory reallocation failed for size list")
+                        sizes = tmp_ptr;
+                        if (NULL ==
+                            (tmp_ptr = H5MM_realloc(vec_bufs, vec_arr_nalloc * sizeof(*vec_bufs) * 2)))
+                            HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
+                                        "memory reallocation failed for buffer list")
+                        vec_bufs = tmp_ptr;
+                    }
+
+                    /* Record that we've doubled the array sizes */
+                    vec_arr_nalloc *= 2;
+                }
+
+                /* Add this segment to vector read list */
+                addrs[vec_arr_nused]    = offsets[i] + file_off[file_seq_i];
+                sizes[vec_arr_nused]    = io_len;
+                vec_bufs[vec_arr_nused] = (void *)((uint8_t *)buf + mem_off[mem_seq_i]);
+                vec_arr_nused++;
+            }
+            else
```

Comment:
I'm looking at this on GitHub and I cannot even get the corresponding `if` onto on the same page at a comfortable type size. This routine needs to be broken down so that a reader can see the program structure and flow one page at a time.